source/darwin/securityframework_rsa.c

```diff
@@ -0,0 +1,545 @@
+        goto on_error;
+    }
+
+    sec_key_export_data = SecKeyCopyExternalRepresentation(key_pair->pub_key_ref, &error);
+    if (error != NULL) {
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        CFRelease(error);
+        goto on_error;
+    }
+
+    key_cur = aws_byte_cursor_from_array(CFDataGetBytePtr(sec_key_export_data), CFDataGetLength(sec_key_export_data));
+
+    if (aws_byte_buf_init_copy_from_cursor(&key_pair->base.pub, allocator, key_cur)) {
+        goto on_error;
+    }
+
+    key_pair->base.vtable = &s_rsa_key_pair_vtable;
+    key_pair->base.key_size_in_bits = key_size_in_bits;
+    key_pair->base.good = true;
+
+    CFRelease(sec_key_export_data);
+    CFRelease(key_size_cf_str);
+    CFRelease(key_attributes);
+
+    return &key_pair->base;
+
+on_error:
+    if (key_attributes) {
+        CFRelease(key_attributes);
+    }
+
+    if (sec_key_export_data) {
+        CFRelease(sec_key_export_data);
+    }
+
+    if (key_size_cf_str) {
+        CFRelease(key_size_cf_str);
+    }
+
+    s_rsa_destroy_key(&key_pair->base);
+    return NULL;
+}
+
+struct aws_rsa_key_pair *aws_rsa_key_pair_new_from_private_key_pkcs1_impl(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor key) {
+    struct sec_rsa_key_pair *key_pair_impl = aws_mem_calloc(allocator, 1, sizeof(struct sec_rsa_key_pair));
+
+    CFMutableDictionaryRef key_attributes = NULL;
+    CFDataRef private_key_data = NULL;
+
+    aws_ref_count_init(&key_pair_impl->base.ref_count, &key_pair_impl->base, aws_rsa_key_pair_destroy);
+    key_pair_impl->base.impl = key_pair_impl;
+    key_pair_impl->base.allocator = allocator;
+    key_pair_impl->cf_allocator = aws_wrapped_cf_allocator_new(allocator);
+    aws_byte_buf_init_copy_from_cursor(&key_pair_impl->base.priv, allocator, key);
+
+    private_key_data = CFDataCreate(key_pair_impl->cf_allocator, key.ptr, key.len);
+
+    if (private_key_data == NULL) {
+        goto on_error;
+    }
+
+    key_attributes = CFDictionaryCreateMutable(key_pair_impl->cf_allocator, 6, NULL, NULL);
+
+    if (key_attributes == NULL) {
+        goto on_error;
+    }
+
+    CFDictionaryAddValue(key_attributes, kSecClass, kSecClassKey);
+    CFDictionaryAddValue(key_attributes, kSecAttrKeyType, kSecAttrKeyTypeRSA);
+    CFDictionaryAddValue(key_attributes, kSecAttrKeyClass, kSecAttrKeyClassPrivate);
+
+    CFErrorRef error = NULL;
+    key_pair_impl->priv_key_ref = SecKeyCreateWithData(private_key_data, key_attributes, &error);
+
+    if (error != NULL) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        CFRelease(error);
+        goto on_error;
+    }
+
+    key_pair_impl->pub_key_ref = SecKeyCopyPublicKey(key_pair_impl->priv_key_ref);
+
+    CFRelease(key_attributes);
+    CFRelease(private_key_data);
+
+    key_pair_impl->base.vtable = &s_rsa_key_pair_vtable;
+    key_pair_impl->base.key_size_in_bits = SecKeyGetBlockSize(key_pair_impl->priv_key_ref) * 8;
```

Comment:
maybe we should just store `key_size_in_bytes` instead of `key_store_in_bits` since it seems like that's what every actual security API is using

if we really want `bits` for our public API, just error-check what the user passes us, and deal with it as bytes from thereon out

or maybe just go with the flow and take bytes like every other API and just make it clear that we expect bytes, not bits like people usually say verbally

Reply:
bits are very typical when talking about RSA. mac, win, lc all take bits as input for key generation. so bits i think make the most sense. 

i'll add an assert on the block size. if everything went correctly we would not be getting sizes that can overflow when multiplied by 8 (it should 4096 max)